### PR TITLE
Add test for shared physical network

### PR DESCRIPTION
**Pull request by hahakiki2010**
Create a new bridge with a physical nic,
and connect it to multiple vms, test if
the vms, host, and remote ip can ping
each other succeed

Signed-off-by: Yan Li <yannli@redhat.com>

### CODE REVIEW

**Comment by hahakiki2010** [comment]
```
(virtenv_py3) [root@localhost libvirt_ci]# avocado run --vt-type libvirt iface_bridge.shared_physical_network 
JOB ID     : 60afe8c1ddf9fd4d1dd9a061ea0a8ebfede699ec
JOB LOG    : /root/avocado/job-results/job-2018-10-22T16.49-60afe8c/job.log
 (1/1) type_specific.io-github-autotest-libvirt.iface_bridge.shared_physical_network: PASS (73.90 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 96.11 s
```

File: libvirt/tests/cfg/virtual_network/iface_bridge.cfg [hunk 1]
```diff
@@ -0,0 +1,18 @@
+- virtual_network.iface_bridge:
+    type = iface_bridge
+    start_vm = no
+    vms = avocado-vt-vm1 vm2
+    bridge_name = "test_br0"
+    # The netdst must be the same with the bridge_name
+    netdst_nic1 = "test_br0"
```
Comment (chunfuwen) [inline libvirt/tests/cfg/virtual_network/iface_bridge.cfg]: use ${bridge_name} instead

File: libvirt/tests/src/virtual_network/iface_bridge.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+        if libvirt.check_iface(bridge_name, "exists", "--all"):
+            virsh.iface_unbridge(bridge_name)
+        if os.path.exists(iface_script_bk):
+            process.run("mv %s %s" % (iface_script_bk, iface_script), shell=True)
+            # Need reload script
+            process.run("ifdown %s" % iface_name, shell=True)
+            process.run("ifup %s" % iface_name, shell=True)
+        if NM_is_running:
+            NM_service.restart()
+        if NW_is_running:
```
Comment (chunfuwen) [inline libvirt/tests/src/virtual_network/iface_bridge.py]: duplicate code

File: libvirt/tests/cfg/virtual_network/iface_bridge.cfg (hunk 1 above)
Comment (hahakiki2010) [reply libvirt/tests/cfg/virtual_network/iface_bridge.cfg]: accepted

File: libvirt/tests/src/virtual_network/iface_bridge.py (hunk 2 above)
Comment (hahakiki2010) [reply libvirt/tests/src/virtual_network/iface_bridge.py]: they are different, NM is for NetworkManager, NW is for network service

File: libvirt/tests/src/virtual_network/iface_bridge.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+    filter_name = params.get("filter_name", "vdsm-no-mac-spoofing")
+    ping_count = params.get("ping_count", "5")
+    ping_timeout = float(params.get("ping_timeout", "10"))
+    iface_name = utils_net.get_net_if(state="UP")[0]
+    iface_script = NETWORK_SCRIPT + iface_name
+    iface_script_bk = os.path.join(data_dir.get_tmp_dir(), "iface-%s.bk" % iface_name)
+
+    vms = params.get("vms").split()
+    if len(vms) <= 1:
+        test.cancel("Need more than two VMs to test")
```
Comment (sathnaga) [inline libvirt/tests/src/virtual_network/iface_bridge.py]: handle cases for more than 2 VMs, or reword this to mention just two VMs.

File: libvirt/tests/src/virtual_network/iface_bridge.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+        NM_service.stop()
+
+    # Start network service
+    # Sometimes network.service not installed by default
+    NW_res = process.run("systemctl status network", shell=True, ignore_status=True)
+    if NW_res.exit_status:
+        if re.search(r'.*(network.service)\s.*not\s.*found', NW_res.stderr_text):
+            if not utils_package.package_install('network-scripts') or \
+                    not utils_package.package_install('initscripts'):
+                test.cancel("Failed to install network service")
```
Comment (sathnaga) [inline libvirt/tests/src/virtual_network/iface_bridge.py]: we should be able to use `service.Factory` itself for this check aswell....can you please figure out, looks like code repeat here...

File: libvirt/tests/src/virtual_network/iface_bridge.py [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+
+        if vm1.is_alive():
+            vm1.destroy()
+
+        vm1.start()
+        # Check if vm can get ip with the new create bridge
+        vm1_ip = vm2_ip = None
+        session1 = session2 = None
+        try:
+            utils_net.update_mac_ip_address(vm1, timeout=240)
```
Comment (sathnaga) [inline libvirt/tests/src/virtual_network/iface_bridge.py]: `vm1.wait_for_login()` should suffice here? as it does all these inside...we can have that on a try block and failout appropriately...

File: libvirt/tests/src/virtual_network/iface_bridge.py [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+            ping(vm1_ip, remote_ip, ping_count, ping_timeout, session=session1)
+
+        # Start vm2 connect to the same bridge
+        modify_iface_xml(bridge_name, filter_name, vm2_name)
+        if vm2.is_alive():
+            vm2.destroy()
+        vm2.start()
+        # Check if vm1 and vm2 can ping each other
+        try:
+            utils_net.update_mac_ip_address(vm2, timeout=240)
```
Comment (sathnaga) [inline libvirt/tests/src/virtual_network/iface_bridge.py]: same as above comment

File: libvirt/tests/src/virtual_network/iface_bridge.py [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+        vm2_xml_bak.sync()
+        virsh.nwfilter_undefine(filter_name, ignore_status=True)
+        if libvirt.check_iface(bridge_name, "exists", "--all"):
+            virsh.iface_unbridge(bridge_name)
+        if os.path.exists(iface_script_bk):
+            process.run("mv %s %s" % (iface_script_bk, iface_script), shell=True)
+            # Need reload script
+            process.run("ifdown %s" % iface_name, shell=True)
+            process.run("ifup %s" % iface_name, shell=True)
+        if NM_is_running:
```
Comment (sathnaga) [inline libvirt/tests/src/virtual_network/iface_bridge.py]: it has been stopped in the teststep, so ideally it should be checked for stopped and invoke start instead restart?

File: libvirt/tests/src/virtual_network/iface_bridge.py (hunk 3 above)
Comment (hahakiki2010) [reply libvirt/tests/src/virtual_network/iface_bridge.py]: updated

File: libvirt/tests/src/virtual_network/iface_bridge.py (hunk 4 above)
Comment (hahakiki2010) [reply libvirt/tests/src/virtual_network/iface_bridge.py]: yes, service.Factory return None when service is not found.  So I guess I can use ```None``` for check here. Thanks for your comments

File: libvirt/tests/src/virtual_network/iface_bridge.py (hunk 5 above)
Comment (hahakiki2010) [reply libvirt/tests/src/virtual_network/iface_bridge.py]: Well, It is just to get the vm ip and do no need to wait vm really get started. But now I think you're right, as vm.wait_for_login is more than the code I picked above. So updated, thanks for your comments

**Comment by hahakiki2010** [comment]
@sathnaga @chunfuwen  all your comments updated. Pls help review again, thanks~

**Comment by kylazhang** [comment]
@hahakiki2010 Please add parameters explanation for all the functions you defined.

File: libvirt/tests/src/virtual_network/iface_bridge.py [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+        vm2.start()
+
+        # Check if vm1 and vm2 can ping each other
+        try:
+            utils_net.update_mac_ip_address(vm2, timeout=120)
+            vm2_ip = vm2.get_address()
+        except Exception as errs:
+            test.fail("vm2 can't get IP with the new create bridge: %s" % errs)
+        session2 = vm2.wait_for_login()
+        if session1 and session2 and vm1_ip and vm2_ip:
```
Comment (kylazhang) [inline libvirt/tests/src/virtual_network/iface_bridge.py]: what to do for else? Or no else then no need to do judgement?

File: libvirt/tests/src/virtual_network/iface_bridge.py [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+        NM_service.stop()
+
+    # Start network service
+    NW_service = service.Factory.create_service("network")
+    NW_status = NW_service.status()
+    if NW_status is None:
+        logging.debug("network service not found")
+        if not utils_package.package_install('network-scripts') or \
+                not utils_package.package_install('initscripts'):
+            test.cancel("Failed to install network service")
```
Comment (kylazhang) [inline libvirt/tests/src/virtual_network/iface_bridge.py]: should be test.error?

File: libvirt/tests/src/virtual_network/iface_bridge.py [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+
+    def create_bridge(br_name, iface_name):
+        """
+        Create a linux bridge by virsh cmd:
+        1. Stop NetworkManager and Start network service
+        2. virsh iface-bridge <iface> <name> [--no-stp]
+        """
+        # Make sure the bridge not exist
+        if libvirt.check_iface(br_name, "exists", "--all"):
+            test.cancel("The bridge %s already exist" % br_name)
```
Comment (kylazhang) [inline libvirt/tests/src/virtual_network/iface_bridge.py]: if exist, why not just use it?

Comment (hahakiki2010) [reply libvirt/tests/src/virtual_network/iface_bridge.py]: Sometimes when bridge already exists, the test env might not be clean, for example the bridge failed to be deleted in last run. So manual recovery would be better here as the scenario is complicated

File: libvirt/tests/src/virtual_network/iface_bridge.py (hunk 9 above)
Comment (hahakiki2010) [reply libvirt/tests/src/virtual_network/iface_bridge.py]: It's a precondition but not a test run failure. So I choose test.cancel. And checked previous cases, half for test.cancel, half for test.error, seemingly both they are accepted.

File: libvirt/tests/src/virtual_network/iface_bridge.py (hunk 8 above)
Comment (hahakiki2010) [reply libvirt/tests/src/virtual_network/iface_bridge.py]: accepted, remove the judgement
